### PR TITLE
Propagate full Unsafe unavailability reason in PlatformDependent

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
It is not clear why Unsafe is unavailable when it is explicitly
disabled, or when Netty thinks it is running on Android.

Modification:
Change the "has" fields and methods to be causes.  A null cause
means Unsafe is present.  This catches all possible reason why
Unsafe might not be available.

Result:
Easier to debug Netty start up when logging cannot be turned on.

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
This is a sibling PR to https://github.com/netty/netty/pull/7679 .   In our environment, turning on logging is sometimes difficult.  Having programmatic access to the exceptions makes it much easier to debug, since we can dump the whole causal chain.


As a side note: I would be interested in making PlatformDependent more testable.  Right now changes to it are based on best guesses of a lot of people.   I think having this well tested would make it easier to refactor when java9 is more prevalent, and breaks lots of things.

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java [hunk 1]
```diff
@@ -369,32 +368,33 @@ public Object run() {
...
         } else {
-            tryUnsafe = SystemPropertyUtil.getBoolean("org.jboss.netty.tryUnsafe", true);
+            unsafePropName = "org.jboss.netty.tryUnsafe";
         }
 
-        if (!tryUnsafe) {
-            logger.debug("sun.misc.Unsafe: unavailable (io.netty.tryUnsafe/org.jboss.netty.tryUnsafe)");
-            return true;
+        if (!SystemPropertyUtil.getBoolean(unsafePropName, true)) {
+            String msg = String.format("sun.misc.Unsafe: unavailable (%s)", unsafePropName);
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: Just use string concatation ? I think there is no much win in use String.format at all.

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 2]
```diff
@@ -965,24 +965,24 @@ private static boolean maybeSuperUser0() {
...
+    private static Throwable unsafeUnavailabilityCause0() {
         if (isAndroid()) {
             logger.debug("sun.misc.Unsafe: unavailable (Android)");
-            return false;
+            return new UnsupportedOperationException("sun.misc.Unsafe: unavailable (Android)");
         }
 
         if (PlatformDependent0.isExplicitNoUnsafe()) {
-            return false;
+            PlatformDependent0.getUnsafeUnavailabilityCause();
```
Comment (trustin) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: Missing `return`, or `return new UnsupportedOperationException("disabled as requested via system property")`?

File: common/src/main/java/io/netty/util/internal/PlatformDependent0.java (hunk 1 above)
Comment (carl-mastrangelo) [reply common/src/main/java/io/netty/util/internal/PlatformDependent0.java]: Done.

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java (hunk 2 above)
Comment (carl-mastrangelo) [reply common/src/main/java/io/netty/util/internal/PlatformDependent.java]: Done.

File: common/src/main/java/io/netty/util/internal/PlatformDependent.java [hunk 3]
```diff
@@ -965,24 +965,24 @@ private static boolean maybeSuperUser0() {
...
 
-    private static boolean hasUnsafe0() {
+    private static Throwable unsafeUnavailabilityCause0() {
         if (isAndroid()) {
             logger.debug("sun.misc.Unsafe: unavailable (Android)");
-            return false;
+            return new UnsupportedOperationException("sun.misc.Unsafe: unavailable (Android)");
         }
 
         if (PlatformDependent0.isExplicitNoUnsafe()) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/PlatformDependent.java]: this conditional is already taken into account in `PlatformDependent0` when setting `PlatformDependent0.getUnsafeUnavailabilityCause()` ... can we just let `PlatformDependent0` be responsible for enforcing this conditional? For example:

```java
Throwable cause = PlatformDependent0.getUnsafeUnavailabilityCause();
if (cause != null) {
  return cause;
}
...
```

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/PlatformDependent.java]: @carl-mastrangelo ^^

Comment (carl-mastrangelo) [reply common/src/main/java/io/netty/util/internal/PlatformDependent.java]: Sorry for the delay, done!

**Comment by normanmaurer** [comment]
@trustin @Scottmitch can you take another look ?

**Comment by normanmaurer** [comment]
@Scottmitch can we merge this ?

**Comment by Scottmitch** [comment]
@normanmaurer - yip ... lgtm sorry for delay

**Comment by normanmaurer** [comment]
@Scottmitch no worries buddy... thanks

**Pull request merged**